staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec_test.go

```diff
@@ -763,6 +794,92 @@ func TestConcurrentUpdateTransportConfig(t *testing.T) {
 	time.Sleep(2 * time.Second)
 }
 
+func TestInstallHintRateLimit(t *testing.T) {
+	c := api.ExecConfig{
+		Command:     "does not exist",
+		APIVersion:  "client.authentication.k8s.io/v1alpha1",
+		InstallHint: "some install hint",
+	}
+	a, err := newAuthenticator(newCache(), &c)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	now := time.Now()
+	r := clientauthentication.Response{}
+
+	tests := []struct {
+		name            string
+		advance         time.Duration
+		calls           int
+		wantInstallHint int
+	}{
+		{
+			name:            "print-up-to-threshold",
+			calls:           a.sometimes.threshold,
+			wantInstallHint: a.sometimes.threshold,
+		},
+		{
+			name:            "dont-print-after-threshold",
+			advance:         a.sometimes.interval / 2,
+			calls:           a.sometimes.threshold,
+			wantInstallHint: 0,
+		},
+		{
+			name:            "after-interval-print-up-to-threshold",
+			advance:         (a.sometimes.interval / 2) + 1,
+			calls:           a.sometimes.threshold,
+			wantInstallHint: a.sometimes.threshold,
+		},
+		{
+			name:            "print-part-of-threshold-part-1",
+			advance:         a.sometimes.interval + 1,
+			calls:           a.sometimes.threshold / 2,
+			wantInstallHint: a.sometimes.threshold / 2,
+		},
+		{
+			name:            "print-part-of-threshold-part-2",
+			advance:         (a.sometimes.interval / 2),
+			calls:           a.sometimes.threshold / 2,
+			wantInstallHint: a.sometimes.threshold / 2,
+		},
+		{
+			name:            "dont-print-part-of-threshold-part-1",
+			advance:         (a.sometimes.interval / 4),
+			calls:           a.sometimes.threshold / 2,
+			wantInstallHint: 0,
+		},
+		{
+			name:            "dont-print-part-of-threshold-part-2",
+			advance:         (a.sometimes.interval / 4) - 1,
+			calls:           a.sometimes.threshold / 2,
+			wantInstallHint: 0,
+		},
+	}
+
+	for _, test := range tests {
```

Comment:
nit: follow the `t.Run` pattern so you can avoid having to pass the test name into `t.Errorf`:

```go
for _, tt := range tests {
	t.Run(tt.name, func(t *testing.T) {
		...
	}
}
```

Reply:
Ok sooooooooo the reason why I didn't use this pattern is because the `tests` depend on one another :(. Which is kinda a smell, but I went with it anyways since I wanted to test the state that `sometimes` holds.

I thought I read in the Kubernetes contribution guidelines that all tests should be able to be run in parallel, which would not be the case for the `TestInstallHintRateLimit/*` subtests since they share state.

Also, if I decided to use `t.Run(...)` here, and someone were to run a single one of the subtests (e.g., `TestInstallHintRateLimit/print-part-of-threshold-part-2`, then it would fail because the previous tests weren't run.

What do you think I should do? Just keep it the way it is?

Reply:
I think if you make it so each test case reproduces the correct state before running the test, it will be easier to understand in addition to being safe for parallel runs. Right now it's not obvious that they have to be run in exact order.

Reply:
Okey dokey, thanks for the help y'all. I gave this a go here: 6ca2b151717002ce3b0dd51f8d0751821879bdb9.